bin/lib/cdn.py

```diff
@@ -0,0 +1,226 @@
+import hashlib
+import tarfile
+import urllib.parse
+
+from base64 import b64encode
+from concurrent import futures
+from datetime import datetime
+from tempfile import mkdtemp
+import os
+import shutil
+import logging
+from pathlib import Path
+
+from lib.amazon import botocore, s3_client, force_lazy_init
+
+logger = logging.getLogger('ce-cdn')
+
+def read_file_chunked(fobj, chunk_size=128*1024):
+    b = bytearray(chunk_size)
+    mv = memoryview(b)
+
+    for n in iter(lambda : fobj.readinto(mv), 0):
+        yield mv[:n]
+
+
+def hash_fileobj(fobj, hash_type):
+    h = hash_type()
+    for chnk in read_file_chunked(fobj):
+        h.update(chnk)
+    return h
+
+
+def hash_file_for_s3(f):
+    with open(f['path'], 'rb') as fobj:
+        sha256 = hash_fileobj(fobj, hashlib.sha256).digest()
+        sha256 = b64encode(sha256).decode()
+        return dict(hash=sha256, **f)
+
+
+def get_directory_contents(basedir):
+    for f in Path(basedir).rglob('*'):
+        name = f.relative_to(basedir).as_posix()
+        yield dict(name=name, path=f)
+
+
+class DeploymentJob(object):
+    tmpdir = None
+
+    def __init__(self, tar_file_path, bucket_name, bucket_path='', version=None, max_workers=None):
+        self.tar_file_path = tar_file_path
+        self.bucket_name = bucket_name
+        self.bucket_path = Path(bucket_path)
+        self.version = version
+        self.max_workers = max_workers or os.cpu_count() or 1
+        self.deploydate = datetime.utcnow().isoformat(timespec='seconds')
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
```

Comment:
Nice :) I've used tmpfile.TemporaryDir() before but this wraps the functionality nicely into the same object. Love it